horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/DatasetServiceImpl.java

```diff
@@ -619,44 +636,40 @@ void createLabelValue(int datasetId, int labelId, JsonNode value) {
     }
 
     @Transactional
-    void createFingerprint(int datasetId, int testId) {
-        JsonNode json = null;
-        try {
-            json = em.createQuery("SELECT t.fingerprintLabels from test t WHERE t.id = ?1", JsonNode.class)
-                    .setParameter(1, testId).getSingleResult();
-        } catch (NoResultException noResultException) {
-            Log.infof("Could not find fingerprint for dataset: %d", datasetId);
-        }
-        if (json == null)
+    void createFingerprint(int datasetId, List<String> fingerprintLabels) {
+        if (fingerprintLabels.isEmpty()) {
             return;
+        }
 
         ObjectNode fpNode = JsonNodeFactory.instance.objectNode();
-        List<LabelValueDAO> labelValues = LabelValueDAO.find("datasetId", datasetId).list();
-        List<String[]> labelPairs = new ArrayList<>(labelValues.size());
-        for (var lv : labelValues)
-            labelPairs.add(new String[] { LabelDAO.<LabelDAO> findById(lv.labelId).name, lv.value.asText() });
-
-        for (int i = 0; i < json.size(); i++)
-            for (var name : labelPairs) {
-                if (json.get(i).asText().equals(name[0]))
-                    fpNode.put(name[0], name[1]);
-            }
+        List<Object[]> filteredLabelValues = session.createNativeQuery(LABEL_VALUES_BY_LABEL_NAMES, Object[].class)
+                .setParameter("datasetId", datasetId)
+                .setParameter("labelNames", fingerprintLabels)
+                .addScalar("name", StandardBasicTypes.TEXT)
+                .addScalar("val", JsonBinaryType.INSTANCE)
+                .getResultList();
+
+        // TODO: can we obtain the json node from the query??
+        for (Object[] lv : filteredLabelValues) {
+            fpNode.put((String) lv[0], ((JsonNode) lv[1]).asText());
+        }
 
         FingerprintDAO fp = new FingerprintDAO();
         fp.datasetId = datasetId;
-        fp.dataset = DatasetDAO.findById(datasetId);
         fp.fingerprint = fpNode;
         fp.populateHash(); // this call is not really necessary
-        if (fp.datasetId > 0 && fp.dataset != null)
+        if (fp.datasetId > 0) {
             fp.persist();
```

Comment:
IN theory you could batch the created `FingerprintDao`s and let hibernate perform a batch insert too
(but I strongly suggest to decide for a specific batch size and probably you need to configure something on hibernate too?).

Reply:
+1 that's very similar to the deletion, I believe we could batch it with proper sizing (technically something we can do on our side w/o additional configuration)